cpp/Platform.Delegates/Platform.Delegates.Delegate.h

```diff
@@ -16,54 +14,134 @@ namespace Platform::Delegates
     template <typename ReturnType, typename... Args>
     class Delegate<ReturnType(Args...)>
     {
+    public:
         using DelegateRawFunctionType = ReturnType(Args...);
         using DelegateFunctionType = std::function<DelegateRawFunctionType>;
 
+        constexpr Delegate() noexcept = default;
+
+        Delegate(const Delegate&) noexcept = default;
+
+        Delegate(Delegate&&) noexcept = default;
+
+        constexpr Delegate(DelegateRawFunctionType simpleFunction) noexcept
+                : simpleFunction(simpleFunction) {}
+
+        Delegate(const DelegateFunctionType &complexFunction)
+                : complexFunction(std::make_shared<DelegateFunctionType>(complexFunction)) {}
+
+        template <typename Class>
+        Delegate(std::shared_ptr<Class> object, ReturnType(Class:: *member)(Args...))
+                : Delegate(std::make_shared<MemberMethod<Class>>(std::move(object), member)) { }
```

Comment:
Hmm.. It is not possible to pass `shared_ptr` directly? Are you sure we have to recreate it?

Reply:
I do not quite understand the question, we create a new shared_ptr here, inside of which is an object

Reply:
Why we need to create a new one, we cannot just use the passed `shared_ptr`?

Reply:
Because it is two different classes. Passed `shared_ptr` is a ptr to `Class`, but we are creating ptr, pointing to the `MemberMethod<Class>`, which contains pointer to `Class` object

Reply:
Wait what? Why we need to create pointer to MemberMethod<Class>? May be we can pass it by value (not by reference)?

Reply:
Ok, looks like I forgot that we are using pointers here.